File: rust-old-capnp/src/all_types.rs
use capnp::{message, serialize, serialize_packed};
use eyre::{Context, Result};
use fuzz_models::{AllTypes, Enum};

use crate::all_types_capnp::{test_all_types, TestEnum};

pub fn read_all_types(src: &[u8], packed: bool) -> Result<AllTypes> {
    let message_reader;
    if packed {
        message_reader =
            serialize_packed::read_message(src, ::capnp::message::ReaderOptions::new())?;
    } else {
        message_reader = serialize::read_message(src, ::capnp::message::ReaderOptions::new())?;
    }

    let mut result = AllTypes::default();

    let at = message_reader
        .get_root::<test_all_types::Reader>()
        .wrap_err("Reader error")?;

    result.bool_field = at.get_bool_field();
    result.int8_field = at.get_int8_field();
    result.int16_field = at.get_int16_field();
    result.int32_field = at.get_int32_field();
    result.int64_field = at.get_int64_field();
    result.uint8_field = at.get_u_int8_field();
    result.uint16_field = at.get_u_int16_field();
    result.uint32_field = at.get_u_int32_field();
    result.uint64_field = at.get_u_int64_field();
    result.float32_field = at.get_float32_field();
    result.float64_field = at.get_float64_field();
    result.enum_field = Enum::from_repr(at.get_enum_field().unwrap() as u8).unwrap();

    if at.has_text_field() {
        result.text_field = Some(at.get_text_field().unwrap().to_string().unwrap())
    }

    Ok(result)
}

fn from_enum(e: Enum) -> TestEnum {
    match (e) {
        Enum::Foo => TestEnum::Foo,
        Enum::Bar => TestEnum::Bar,
        Enum::Baz => TestEnum::Baz,
        Enum::Qux => TestEnum::Qux,
        Enum::Quux => TestEnum::Quux,
        Enum::Corge => TestEnum::Corge,
        Enum::Grault => TestEnum::Grault,
        Enum::Garply => TestEnum::Garply,
    }
}

fn serialize_all_types_helper(v: &AllTypes, builder: &mut test_all_types::Builder<'_>) {
    builder.set_bool_field(v.bool_field);
    builder.set_int8_field(v.int8_field);
    builder.set_int16_field(v.int16_field);
    builder.set_int32_field(v.int32_field);
    builder.set_int64_field(v.int64_field);
    builder.set_u_int8_field(v.uint8_field);
    builder.set_u_int16_field(v.uint16_field);
    builder.set_u_int32_field(v.uint32_field);
    builder.set_u_int64_field(v.uint64_field);
    builder.set_float32_field(v.float32_field);
    builder.set_float64_field(v.float64_field);
    builder.set_enum_field(from_enum(v.enum_field));

    if let Some(text_field) = v.text_field.as_ref() {
        builder.set_text_field(text_field);
    }
}

pub fn serialize_all_types(v: &AllTypes, packed: bool) -> Result<Vec<u8>> {
    let mut target = vec![];
    let mut message = ::capnp::message::Builder::new_default();
    {
        let mut all_types = message.init_root::<test_all_types::Builder>();
        serialize_all_types_helper(v, &mut all_types);
    }

    if packed {
        serialize_packed::write_message(&mut target, &message).wrap_err("Serialization error")?;
    } else {
        serialize::write_message(&mut target, &message).wrap_err("Serialization error")?;
    }

    Ok(target)
}

File: c++/all_types.cpp

#include "all_types.capnp.h"
#include <capnp/message.h>
#include <capnp/serialize-packed.h>
#include <kj/io.h>
#include <iostream>
#include <vector>
#include <string>
#include <stdexcept>
#include "recapn_bench_cpp.h"
#include "../fuzz-models/fuzz_models.h"

using namespace std;

TestEnum from_enum(Enum raw)
{
    switch (raw)
    {
    case Enum::Foo:
        return TestEnum::FOO;
    case Enum::Bar:
        return TestEnum::BAR;
    case Enum::Baz:
        return TestEnum::BAZ;
    case Enum::Qux:
        return TestEnum::QUX;
    case Enum::Quux:
        return TestEnum::QUUX;
    case Enum::Corge:
        return TestEnum::CORGE;
    case Enum::Grault:
        return TestEnum::GRAULT;
    case Enum::Garply:
        return TestEnum::GARPLY;
    }
}

void write_all_types_helper(
    TestAllTypes::Builder &target,
    AllTypesC *v,
    kj::OutputStream &output)
{
    target.setBoolField(v->bool_field);
    target.setInt8Field(v->int8_field);
    target.setInt16Field(v->int16_field);
    target.setInt32Field(v->int32_field);
    target.setInt64Field(v->int64_field);
    target.setUInt8Field(v->uint8_field);
    target.setUInt16Field(v->uint16_field);
    target.setUInt32Field(v->uint32_field);
    target.setUInt64Field(v->uint64_field);
    target.setFloat32Field(v->float32_field);
    target.setFloat64Field(v->float64_field);
    target.setEnumField(from_enum(v->enum_field));

    if (v->text_field.present)
    {
        target.setTextField(capnp::Text::Reader(
            (char *)v->text_field.value.start,
            (size_t)v->text_field.value.size));
    }

    if (v->data_field.present)
    {
        size_t size = (size_t)v->data_field.value.size;
        if (size > 0)
        {
            target.setDataField(capnp::Data::Reader(
                (u_int8_t *)v->data_field.value.start,
                size));
        }
        else
        {
            target.setDataField(kj::Array<u_int8_t>(0));
        }
    }
}

void write_all_types_to_buffer(AllTypesC *v, kj::OutputStream &output, bool packed)
{
    ::capnp::MallocMessageBuilder message;

    TestAllTypes::Builder root = message.initRoot<TestAllTypes>();

    write_all_types_helper(root, v, output);

    if (packed)
    {
        writePackedMessage(output, message);
    }
    else
    {
        writeMessage(output, message);
    }
}

Buffer serialize_all_types(AllTypesC *v, bool packed)
{

    Buffer result;
    result.kj_array_buffer = (void *)0;

    try
    {
        kj::VectorOutputStream *stream = new kj::VectorOutputStream(1024);

        write_all_types_to_buffer(v, *stream, packed);

        result.kj_array_buffer = (void *)stream;

        result.size = stream->getArray().asBytes().size();
    }
    catch (const std::exception &e)
    {
        // catch anything thrown within try block that derives from std::exception
        std::cerr << e.what();
    }

    return result;
}